File: ds/doubly_linked_list.py
class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None

class LinkedList:
    def __init__(self):
        self.head = ListNode(-1)
        self.tail = ListNode(-1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def insertFront(self, val):
        newNode = ListNode(val)
        newNode.prev = self.head
        newNode.next = self.head.next

        self.head.next.prev = newNode
        self.head.next = newNode

    def insertEnd(self, val):
        newNode = ListNode(val)
        newNode.next = self.tail
        newNode.prev = self.tail.prev
        self.tail.prev.next = newNode
        self.tail.prev = newNode

    def removeFront(self):
        self.head.next.next.prev = self.head
        self.head.next = self.head.next.next

    def removeEnd(self):
        self.tail.prev.prev.next = self.tail
        self.tail.prev = self.tail.prev.prev

    def print(self):
        curr = self.head.next
        while curr != self.tail:
            print(curr.val)
            curr = curr.next

# test cases
def test_linked_list():
    ll = LinkedList()
    ll.insertEnd(1)
    ll.insertFront(13)
    ll.insertFront(2)
    ll.insertEnd(20)

    ll.removeEnd()
    ll.removeFront()

    assert ll.head.next.val == 13
    assert ll.tail.prev.val == 1

    print('All test cases pass')

test_linked_list()

File: ds/doubly_linked_list.js
class ListNode {
    constructor(val){
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

class LinkedList {
    constructor(){
        this.head = new ListNode(-1);
        this.tail = new ListNode(-1);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    insertFront(val){
        let newNode = new ListNode(val);
        newNode.next = this.head.next;
        newNode.prev = this.head;
        this.head.next.prev = newNode;
        this.head.next = newNode;
    }

    insertEnd(val){
        let newNode = new ListNode(val);
        newNode.next = this.tail;
        newNode.prev = this.tail.prev;

        this.tail.prev.next = newNode;
        this.tail.prev = newNode;
    }

    removeFront(){
        this.head.next.next.prev = this.head;
        this.head.next = this.head.next.next;
    }

    removeEnd(){
        let node = this.tail.prev;
        node.prev.next = this.tail;
        this.tail.prev = node.prev;
    }
}

function testLinkedList(){
    const assert = require('assert');

    let linkedList = new LinkedList();
    linkedList.insertEnd(1);
    linkedList.insertEnd(2);
    linkedList.insertEnd(3);
    linkedList.insertEnd(4);
    linkedList.insertEnd(5);

    linkedList.removeEnd();
    assert(linkedList.tail.prev.val === 4);

    linkedList.removeFront();
    assert(linkedList.head.next.val === 2);

    linkedList.insertFront(0);
    assert(linkedList.head.next.val === 0);

    linkedList.removeFront();
    assert(linkedList.head.next.val === 2);

    linkedList.removeEnd();
    assert(linkedList.tail.prev.val === 3);

    console.log('All test cases pass');
}

testLinkedList();